File: 04/p4.rb
# Find the largest palindome number (9009) that
# is the product of 2 3-digit numbers.

def to_pal(n)
  (n.to_s + n.to_s.reverse).to_i
end

def pal_product?(n)
  (100..1000).any? do |i|
    (n % i == 0) && (n / i) < 1000
  end
end

range = (1..1000).reverse_each
match = range.find { |i| pal_product?(to_pal(i)) }
large_pal = to_pal(match)

puts large_pal
# 906609
# 913 * 993

File: 04/p4.jl
# Find the largest palindome number (9009) that is the product of 2 3-digit numbers.

to_pal(n) = int("$n$(reverse("$n"))")

function product_of_2_3s(n)
  for i = 100:999
    if n % i == 0 && n / i < 999
      return true
    end
  end
  false
end

function large_pal()
  i = 999
  while !product_of_2_3s(to_pal(i))
    i -= 1
  end
  to_pal(i)
end

println(large_pal())
# 906609
# 913 * 993
